### DISCUSSION TITLE
bash: bleopt/check:import_path: No such file or directory

### DISCUSSION THREAD

**Discussion by Freed-Wu** [akinomyoga/ble.sh#465] [Q&A] [answered]
```sh
bleopt import_path="${XDG_DATA_HOME:-$HOME/.local/share}/blesh/local:/usr/share/blesh/contrib"
```

It will give me:

```
bash: bleopt/check:import_path: No such file or directory
```

What happened?

**Comment by akinomyoga** [comment]
One possibility is that you use an old version of ble.sh with the POSIX mode of the unreleased Bash 5.3. What is the output of the following command?

```bash
$ echo "$SHELLOPTS"
$ (set +o posix; ble/widget/display-shell-version)
```

**Reply by Freed-Wu** [reply]
```sh
$ echo "$SHELLOPTS"
braceexpand:emacs:history:interactive-comments:monitor
$ (set +o posix; ble/widget/display-shell-version)
bash: x/run/current-system/sw/etc/bash_completion.d: division by 0 (error token is "run/current-system/sw/etc/bash_completion.d")

[ble: exit 1]
$ ble/widget/display-shell-version
bash: x/run/current-system/sw/etc/bash_completion.d: division by 0 (error token is "run/current-system/sw/etc/bash_completion.d")
```

**Reply by akinomyoga** [reply]
Thanks. Hmm, this is strange.

- What is your environment?
- What is the Bash version?
- What is the version of ble.sh?
- What are the outputs of the following commands?

```bash
$ builtin type builtin
$ builtin type ble/is-function
$ builtin type declare
```

**Reply by Freed-Wu** [reply]
Oh, I have an function `declare` to `declare -f function_name | bat -pplbash`. Is it because ble.sh don't allow it?

**Reply by akinomyoga** [reply]
Could you answer the questions?

**Reply by Freed-Wu** [reply]
Oh, sorry! After removing `declare`, everything works normally.

**Reply by akinomyoga** [reply]
> Oh, I have an function `declare` to `declare -f function_name | bat -pplbash`.

`declare` is one of the basic builtin commands of Bash that constitutes the Bash language. If you change the meaning of the `declare` builtin, it means you effectively define a language different from the Bash language. Then, you cannot expect any "Bash" scripts and functions to work properly.

> Is it because ble.sh don't allow it?

Actually, `ble.sh` allows users to override the `declare` command, (or otherwise, ble.sh wouldn't have worked in the first place). However, `ble.sh` still runs the user commands (i.e. the commands that the user inputs and executes through the command line) with the custom `declare` command, so the command `bleopt` executed from the command line is processed using the custom `declare` that you define.

**Reply by akinomyoga** [reply]
> Oh, sorry! After removing `declare`, everything works normally.

OK, I'll summarize the problem and solution to this discussion in an independent reply.

**Reply by akinomyoga** [reply]
@Freed-Wu I made https://github.com/akinomyoga/ble.sh/discussions/465#discussioncomment-9975196 so that you can mark it as an "answer" for this Q\&A.

**Comment by akinomyoga** [comment, marked as answer]
This is the summary.

If one expects the commands executed through the command line to work properly, one shouldn't override the builtin commands of Bash to have a different meaning. In the reported case, the `declare` builtin was overridden. If one wants to make shell functions using the builtin work properly, one needs to unset the overriding functions/aliases/etc. and recover the original builtins.

Note: Even if you try to override builtins without changing their behavior, it is impossible to emulate the behavior of `builtin`, `eval`, `declare`, `type`, `local`, `unset`, `exec`, `return`, `break`, `continue`, `trap`, etc. Their behavior changes if they are called inside a shell function indirectly.
